cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusEventPublisher.java

```diff
@@ -163,41 +167,54 @@ protected void processMessages(StructuredTableContext structuredTableContext,
         .withUserArgs(GSON.fromJson(userArgsString, argsMapType))
         .withSystemArgs(GSON.fromJson(sysArgsString, argsMapType));
       if (programRunStatus.isEndState()) {
-        populateErrorDetailsAndMetrics(builder, properties, programRunStatus, programRunId, publishTime);
+        populateErrorDetailsAndMetrics(builder, properties, programRunStatus, programRunId, notifiedTime);
       } else {
-        writeInEventWriters(builder, publishTime);
+        writeInEventWriters(builder, notifiedTime);
       }
     });
   }
 
-  private void writeInEventWriters(ProgramStatusEventDetails.Builder builder, long publishTime) {
+  private void writeInEventWriters(ProgramStatusEventDetails.Builder builder, long notifiedTime) {
+    long publishTime = System.currentTimeMillis();
     ProgramStatusEventDetails programStatusEventDetails = builder.build();
     ProgramStatusEvent programStatusEvent = new ProgramStatusEvent(publishTime, EVENT_VERSION,
                                                                    instanceName,
                                                                    projectName, programStatusEventDetails);
     List<ProgramStatusEvent> listProgramStatus = new ArrayList<>();
     listProgramStatus.add(programStatusEvent);
+    emitMetrics(notifiedTime, publishTime, programStatusEvent);
     this.eventWriters.forEach(eventWriter -> eventWriter.write(listProgramStatus));
   }
 
+  private void emitMetrics(long notifiedTime, long publishTime, ProgramStatusEvent programStatusEvent) {
+    Map<String, String> metricTags = new HashMap<>();
+    metricTags.put(Constants.Metrics.Tag.NAMESPACE, programStatusEvent.getEventDetails().getNamespace());
+    metricTags.put(Constants.Metrics.Tag.STATUS, programStatusEvent.getEventDetails().getStatus());
+    metricTags.put(Constants.Metrics.Tag.PROGRAM, programStatusEvent.getEventDetails().getProgramName());
+    metricTags.put(Constants.Metrics.Tag.APP, programStatusEvent.getEventDetails().getApplicationName());
+    metricsCollectionService.getContext(metricTags).increment(Constants.Metrics.ProgramEvent.PUBLISHED_COUNT, 1L);
+    metricsCollectionService.getContext(metricTags)
```

Comment:
Can we add a comment here explaining that the latency is mostly only applicable if the status is in completed state? Otherwise we may end up in the situation in which other non-completed program statuses skew the overall latencies downwards.

Reply:
Moved this to SparkProgramStatusMetricsProvider and changed it to a metric specific to metrics fetching.